### PR TITLE
[HEP] [Dataset] Custom Form generator script

### PR DESCRIPTION
**Pull request by matiasgarcia91**
### :pushpin: References
* **Issue:** close #6 

### :tophat: What is the goal?

Create script to generate custom forms for Hepatitis Cascade DataEntry

### :memo: Implementation
Script creates a custom form composed of:
- One tab per dataSet section with a Table containing: 
1. One row per `INTEGER_ZERO_OR_POSITIVE`, `PERCENTAGE` DataElements
2. Two columns per categoryOptionCombo (`HBV`, `HCV`), one with the fields for previous row DataElements input, and a second one with a checkbox group titled "Source of Data". All `TRUE_ONLY` DataElements are added to this group.
- Outside of the tab structure one "Additional Info field" per DataElement with `GLOBAL_ENTRY` code and per categoryOptionCombo is included.

### :art: Screenshots
image

### :boom: How can it be tested?
`cd` into HEP folder and follow README

### :floppy_disk: Requires DB migration?

- [x] Nope, we can just merge this branch.
- [ ] Yes, but we need to apply it before merging this branch.
- [ ] Yes, it's already applied.

### CODE REVIEW

File: HEP/src/components/GlobalTextFields.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+interface GlobalTextFieldAttributes {
+    dataElements: SectionDataElement[];
+    categoryOptionCombos: CategoryOptionCombo[];
+}
+
+export function GlobalTextFields(attributes: GlobalTextFieldAttributes): string {
+    const { dataElements, categoryOptionCombos } = attributes;
+    const fields = dataElements.map(de =>
+        categoryOptionCombos.map(coc => [
```
Comment (tokland) [inline HEP/src/components/GlobalTextFields.tsx]: `map` inside a `map`, this creates an array of arrays. We can do a _.flatMap or _.flatten just before using it (now I see that's what you are doing in the rest of the code, so this one)

File: HEP/src/models/Form.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    public static getFormHtml(dataSet: DataSet): string {
+        const sections = this.getOrderedSections(dataSet.sections);
+        const formHtml = CustomForm({ sections });
+        return formHtml;
+    }
+
+    public static getCategoryOptionCombos(
+        dataElements: SectionDataElement[]
+    ): CategoryOptionCombo[] {
+        return dataElements[0].categoryCombo.categoryOptionCombos;
```
Comment (tokland) [inline HEP/src/models/Form.ts]: Same than before, check on `dataElements[0]`.

File: HEP/src/models/Form.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            } else if (valueTypes.entryField.includes(dataElement.valueType)) {
+                const fields = [...acc.fields, dataElement];
+                return { ...acc, fields };
+            } else {
+                const checkboxes = [...acc.checkboxes, dataElement];
+                return { ...acc, checkboxes };
+            }
+        },
+        { checkboxes: [], fields: [], globalEntry: [] }
+    );
```
Comment (tokland) [inline HEP/src/models/Form.ts]: `reduce` is the universal iterator for functional code, but same than the imperative `each`, it should be used only if higher abstractions do not work. Here the common structure of the if blocks signals you are doing a grouping. Proposal (untested):

```
function sortDataElements(dataElements: SectionDataElement[]): OrderedSection {
    const { globalEntry, fields, checkboxes } = _.groupBy(dataElements, dataElement => {
        if (valueTypes.globalEntryCode === dataElement.code) {
            return "globalEntry";
        } else if (valueTypes.entryField.includes(dataElement.valueType)) {
            return "fields";
        } else {
            return "checkboxes";
        }
    });

    return { globalEntry, fields, checkboxes };
}
```

File: HEP/src/cli.ts [hunk 4]
```diff
@@ -37,52 +32,54 @@ async function getDataSetPayload(
     dataSetId: string
 ): Promise<MetadataPayload> {
```
Comment (tokland) [inline HEP/src/cli.ts]: This scripts is now pretty simple and it's ok to have this `getDataSetPayload`. Whenever it grows a bit more, I'd move all this logic somewhere else and leave cli.js as simple as possible, just the processing of arguments and single calls to methods that do all the work.

File: HEP/src/components/GlobalTextFields.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+            <div class="global-entry-title">{`${de.formName} for ${coc.name}`}</div>,
+            <textarea
+                id={`${de.id}-${coc.id}-val`}
+                name="entryfield"
+                class="entryfield entryarea custom-text-area"
+                placeholder="Add your comment here"
```
[the remark this answers is not in the document]
Comment (matiasgarcia91) [reply HEP/src/components/GlobalTextFields.tsx]: @ifoche what do you think about this? are translations necessary?

File: HEP/src/cli.ts [hunk 6]
```diff
@@ -37,52 +32,54 @@ async function getDataSetPayload(
...
-async function getCustomFormHtml(): Promise<string> {
-    const html = CustomFormExample({});
+async function getAssembledHtml(formHtml: string, dataSetId: string): Promise<string> {
     const style = await getResource("custom-form.css");
     const javascript = await getResource("custom-form.js");
     const styleHtml = `<style>${style}</style>`;
-    const javascriptHtml = `<script type="text/javascript">${javascript}</script>`;
-    return styleHtml + javascriptHtml + html;
+    const javascriptHtml = `<script id="custom-form-script" type="text/javascript" dataSetId=${dataSetId}>${javascript}</script>`;
+    const fontAwesome = `<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.0/css/all.css" integrity="sha384-lZN37f5QGtY3VHgisS14W3ExzMWZxybE1SJSEsQp9S+oqd12jhcu+A56Ebc1zFSJ" crossorigin="anonymous">`;
```
[the remark this answers is not in the document]
Comment (tokland) [reply HEP/src/cli.ts]: Can we use `AssembledFormHTML` like in HEP-data-entry?

**Pull request merged**